File: server/src/main/java/com/smartexpense/api/infrastructure/persistence/mapper/NotificationMapper.java
package com.smartexpense.api.infrastructure.persistence.mapper;

import com.smartexpense.api.domain.model.Notification;
import com.smartexpense.api.infrastructure.persistence.entity.NotificationEntity;
import com.smartexpense.api.infrastructure.persistence.entity.UserEntity;
import org.springframework.stereotype.Component;

@Component
public class NotificationMapper {

    public Notification toDomain(NotificationEntity entity) {
        if (entity == null) {
            return null;
        }
        return Notification.builder()
                .id(entity.getId())
                .userId(entity.getUser().getId())
                .type(entity.getType())
                .content(entity.getContent())
                .isRead(entity.getIsRead())
                .createdAt(entity.getCreatedAt())
                .build();
    }

    public NotificationEntity toEntity(Notification domain) {
        if (domain == null) {
            return null;
        }
        UserEntity user = new UserEntity();
        user.setId(domain.getUserId());

        return NotificationEntity.builder()
                .id(domain.getId())
                .user(user)
                .type(domain.getType())
                .content(domain.getContent())
                .isRead(domain.getIsRead())
                .createdAt(domain.getCreatedAt())
                .build();
    }
}

File: server/bin/main/com/smartexpense/api/infrastructure/persistence/mapper/NotificationMapper.kt
package com.smartexpense.api.infrastructure.persistence.mapper

import com.smartexpense.api.domain.model.Notification
import com.smartexpense.api.infrastructure.persistence.entity.NotificationEntity
import com.smartexpense.api.infrastructure.persistence.entity.UserEntity
import org.springframework.stereotype.Component

@Component
class NotificationMapper {

    fun toDomain(entity: NotificationEntity) = Notification(
        id = entity.id,
        userId = entity.user.id!!,
        type = entity.type,
        content = entity.content,
        isRead = entity.isRead,
        createdAt = entity.createdAt
    )

    fun toEntity(domain: Notification): NotificationEntity {
        // Create a proxy UserEntity with only the ID to satisfy the FK relationship
        val userRef = UserEntity(
            id = domain.userId,
            email = "",
            passwordHash = "",
            displayName = ""
        )
        return NotificationEntity(
            id = domain.id,
            user = userRef,
            type = domain.type,
            content = domain.content,
            isRead = domain.isRead,
            createdAt = domain.createdAt
        )
    }
}
